deeppavlov/models/syntax_parser/network.py

```diff
@@ -0,0 +1,345 @@
+        super().__init__(keep_prob=keep_prob,
+                         bert_config_file=bert_config_file,
+                         pretrained_bert=pretrained_bert,
+                         attention_probs_keep_prob=attention_probs_keep_prob,
+                         hidden_keep_prob=hidden_keep_prob,
+                         encoder_layer_ids=encoder_layer_ids,
+                         encoder_dropout=encoder_dropout,
+                         optimizer=optimizer,
+                         weight_decay_rate=weight_decay_rate,
+                         ema_decay=ema_decay,
+                         ema_variables_on_cpu=ema_variables_on_cpu,
+                         freeze_embeddings=freeze_embeddings,
+                         learning_rate=learning_rate,
+                         bert_learning_rate=bert_learning_rate,
+                         min_learning_rate=min_learning_rate,
+                         learning_rate_drop_div=learning_rate_drop_div,
+                         learning_rate_drop_patience=learning_rate_drop_patience,
+                         load_before_drop=load_before_drop,
+                         clip_norm=clip_norm,
+                         **kwargs)
+
+    def _init_graph(self) -> None:
+        self._init_placeholders()
+
+        units = super()._init_graph()
+
+        with tf.variable_scope('ner'):
+            units = token_from_subtoken(units, self.y_masks_ph)
+            if self.use_birnn:
+                units, _ = bi_rnn(units,
+                                  self.birnn_hidden_size,
+                                  cell_type=self.birnn_cell_type,
+                                  seq_lengths=self.seq_lengths,
+                                  name='birnn')
+                units = tf.concat(units, -1)
+            # for heads
+            head_embeddings = tf.layers.dense(units, units=self.state_size, activation="relu")
+            head_embeddings = tf.nn.dropout(head_embeddings, self.embeddings_keep_prob_ph)
+            dep_embeddings = tf.layers.dense(units, units=self.state_size, activation="relu")
+            dep_embeddings = tf.nn.dropout(dep_embeddings, self.embeddings_keep_prob_ph)
+            self.dep_head_similarities = biaffine_attention(dep_embeddings, head_embeddings)
+            self.dep_heads = tf.argmax(self.dep_head_similarities, -1)
+            self.dep_head_probs = tf.nn.softmax(self.dep_head_similarities)
+            # for dependency types
+            head_embeddings = tf.layers.dense(units, units=self.state_size, activation="relu")
+            head_embeddings = tf.nn.dropout(head_embeddings, self.embeddings_keep_prob_ph)
+            dep_embeddings = tf.layers.dense(units, units=self.state_size, activation="relu")
+            dep_embeddings = tf.nn.dropout(dep_embeddings, self.embeddings_keep_prob_ph)
+            # matching each word with its head
+            head_embeddings = gather_indexes(head_embeddings, self.y_head_ph)
```

Comment:
head_embeddings and dep_embeddings variable names clash confuses me. Maybe it is better to use `deepcopy(dep_embeddings)` & `deepcopy(head_embeddings)`?

Reply:
But why do I need `deepcopy`? There is no copying there, I just reuse the same name.